frappe/model/sync.py

```diff
@@ -135,3 +136,89 @@ def get_doc_files(files, start_path):
 							files.append(doc_path)
 
 	return files
+
+
+def remove_stale_doctypes():
+	"""Find and remove any stale doctypes."""
+
+	doctype_names = frappe.get_all("DocType", {"istable": 0, "custom": 0}, pluck="name")
+	stale_doctype = []
+
+	for doctype in doctype_names:
+		try:
+			get_controller(doctype=doctype)
+		except ImportError:
+			stale_doctype.append(doctype)
+
+	length = len(stale_doctype)
+	if length == 0:
+		return
+
+	print(f"{length} stale DocType/s found.")
+	for i, name in enumerate(stale_doctype):
+		frappe.delete_doc("DocType", name, force=True, ignore_missing=True)
+		frappe.db.commit()
+
+		update_progress_bar("Deleting non-existant DocTypes", i, length)
+	print()
+
+
+def remove_stale_reports():
+	"""Find and remove any stale reports."""
+
+	if not frappe.get_site_config().get("developer_mode"):
+		# Skip if developer mode is not enabled.
+		print("Enable developer mode to check for stale reports.")
```

Comment:
Is it risky to run this in prod? it only touches non-standard reports.

Reply:
- If we look at the `on_trash` method for report DOCTYPE, we are not allowed to delete reports if we are not in `developer mode`  or running a `patch`.
https://github.com/frappe/frappe/blob/develop/frappe/core/doctype/report/report.py#L85
- If we try to delete reports in production, validation exception will be raised and migration will fail.